[4-9단계 - 방탈출 예약 관리] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [ ] 4
- [x] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
객체지향 생활체조 요구사항을 잘 충족하였다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

코드 가독성 측면과 무의식적으로 사용했을만한 부분에 집중해서 리뷰 부탁드립니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

RaZel713 committed:
init: 초기 연결 설정

RaZel713 committed:
docs: step1 README 작성

RaZel713 committed:
feat: 어드민 페이지 응답 구현

changuii committed:
feat: 웰컴 페이지 응답 구현

changuii committed:
docs: 2단계 요구사항 정의

changuii committed:
feat: admin/reservation 요청시 예약 페이지 응답 구현

RaZel713 committed:
feat: 예약 조회 기능 구현

changuii committed:
docs: 3단계 미션 요구사항 작성

changuii committed:
feat: 예약 생성 기능 구현

RaZel713 committed:
feat: 예약 삭제 기능 구현

changuii committed:
refactor: Admin 페이지는 AdminCotroller에 작성

changuii committed:
refactor: ReservationController 리팩터링

changuii committed:
refactor: 패키지 분리

changuii committed:
test: 테스트 패키지 분리

changuii committed:
fix: getter()가 없어 reservation이 json으로 잘못 변환되는 문제 해결

changuii committed:
refactor: final 키워드 추가

RaZel713 committed:
docs: 요구 사항만 작성하도록 변경

RaZel713 committed:
refactor: ReservationController 의 메서드 명을 더 자세히 작성

RaZel713 committed:
refactor: ReservationController의 메서드 순서 변경

- POST, GET, PUT, PATCH, DELETE 순으로 정렬

RaZel713 committed:
refactor: 예약 로직을 Reservations 객체가 관리

RaZel713 committed:
feat: 주어진 아이디에 해당하는 예약이 없다면 404로 응답

changuii committed:
test: 예약 api들을 분리해서 테스트 작성

- 생성, 조회, 삭제 기능들을 각각 분리해서 호출 테스트

changuii committed:
test: api 호출 테스트를 위한 클래스 이름을 명확하게 변경

changuii committed:
chore: 클래스 선언과 첫 필드 변수 사이 개행 처리

RaZel713 committed:
refactor: Reservation을 불변 객체로 변환

RaZel713 committed:
refactor: 네이밍 변경 및 불필요한 코드 제거

changuii committed:
feat: id를 작성하는 의도를 명확하게 전달하라

- 현재 Reservation id가 비어있거나, write하려는 id가 비어있다면 예외가 발생한다.

changuii committed:
refactor: Reservations의 메서드 명을 더 명확하게 작성

changuii committed:
test: WebEnvinoment.RANDOM_PORT로 수정

- RestAssured에도, port를 설정해주었다.

changuii committed:
docs: 4단계 요구사항 정의

changuii committed:
feat: application.properties에 h2 데이터베이스 설정

changuii committed:
feat: 데이터베이스 초기화 schema.sql 추가

changuii committed:
test: 4단계 미션 테스트 추가

changuii committed:
docs: 5단계 요구사항 정의

changuii committed:
test: 5단계 테스트 추가

changuii committed:
feat: reservations 조회 기능을 데이터베이스에서 조회하도록 변경

changuii committed:
docs: 6단계 요구사항 추가

changuii committed:
refactor: reservation을 데이터베이스에 저장하도록 변경

changuii committed:
refactor: 삭제 기능을 데이터베이스에서 삭제하도록 변경

changuii committed:
chore: Spring Boot 배너 변경

changuii committed:
test: 6단계 테스트 추가

changuii committed:
test: 테스트를 모두 생성자 주입으로 변경

changuii committed:
remove: 사용하지 않는 클래스 제거

changuii committed:
feat: DTO 도입

changuii committed:
docs: 7단계 요구사항 정의

changuii committed:
feat: time 테이블 정의

changuii committed:
feat: 7단계 테스트 정의

changuii committed:
refactor: keyHolder가 아닌, SimpleJdbcInsert를 사용해 insert하도록 변경

- 가독성 향상

changuii committed:
refactor: reservation을 새로 생성자 대신 writeId를 사용

changuii committed:
feat: Time 생성 구현

changuii committed:
feat: Time 모두 조회 구현

changuii committed:
feat: Time 삭제 기능 추가

changuii committed:
docs: 8단계 요구사항 정의

changuii committed:
feat: 레거시 페이지를 새로운 예약 페이지로 변경

changuii committed:
feat: Reservation과 Time을 연결

changuii committed:
feat: Time의 id를 조회할 때, 값이 아직 존재하는지 검증 로직 추가

changuii committed:
refactor: TimeDao에게 의존하지 않고 ReservationDao에서 모든 값을 조회하도록 변경

changuii committed:
test: 8단계 테스트 추가

changuii committed:
test: /admin/time 페이지 테스트

changuii committed:
chore: 사용하지 않는 코드 제거

changuii committed:
refactor: Dao는 id를 반환하고, service에서 값을 찾아서 반환하도록 변경

changuii committed:
chore: 사용하지 않는 메서드 제거

changuii committed:
test: TimeService 테스트

changuii committed:
refactor: ReservationJdbcDao 리팩터링

changuii committed:
refactor: RestController로 명시

changuii committed:
refactor: TimeDao 리팩터링

changuii committed:
docs: 9단계 요구사항 정의

changuii committed:
test: 9단계 테스트 추가

changuii committed:
fix: IllegalArgumentException을 잡도록 변경

changuii committed:
test: TimeApiTest 추가

changuii committed:
test: 테스트 준비 단계의 로직을 메서드로 추출

changuii committed:
test: port가 변하지 않도록 final 키워드 추가

changuii committed:
test: 읽기 쉽도록 주석 추가

changuii committed:
test: 읽기 쉽도록 주석 추가

changuii committed:
test: 읽기 쉽도록 주석 추가

changuii committed:
chore: dto 패키지 추가

changuii committed:
Merge branch 'changuii' into step2

changuii committed:
fix: 잘못 merge된 내역 수정

changuii commented on src/main/java/roomescape/reservation/dto/ReservationRequest.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.LocalDate;
+import java.util.Objects;
+
+public record ReservationRequest(
+        String name,
+        LocalDate date,
+        Long timeId
+) {
+    public ReservationRequest {
+        Objects.requireNonNull(name);
+        Objects.requireNonNull(date);
+        Objects.requireNonNull(timeId);
+    }
+}
```
Request라는 객체를 만든 이유는 api 클라이언트 입장에서 필요한 값들을 정의하기 위해 만들었습니다.
- 기존에는 Reservation을 그대로 사용했지만 실제 api 요청에는 `id`가 필요하지 않기 때문에 이를 명확하게 하고, Reservation이 api 요청 요구사항의 변경에 영향을 받지 않도록 하기 위해 이렇게 분리했습니다.

changuii commented on src/main/java/roomescape/reservation/dto/ReservationResponse.java:
```diff
@@ -0,0 +1,22 @@
+import roomescape.reservation.Reservation;
+import roomescape.time.dto.TimeResponse;
+
+public record ReservationResponse(
+        long id,
+        String name,
+        LocalDate date,
+        TimeResponse time
+) {
+
+    public static ReservationResponse createResponse(final Reservation reservation) {
+        return new ReservationResponse(
+                reservation.id(),
+                reservation.name(),
+                reservation.date(),
+                TimeResponse.createResponse(reservation.time())
+        );
+    }
+}
```
ReservationResponse도 ReservationRequest와 동일하게, 응답해줄 값의 변경이 Reservation에 영향을 주지 않기 위해서 만들었습니다.

changuii commented on src/main/java/roomescape/reservation/ReservationDao.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public interface ReservationDao {
+    Long saveReservation(Reservation reservation, Long timeId);
+    List<Reservation> findAllReservation();
+    Reservation findReservationById(Long id);
+    void deleteReservationById(long id);
+}
```
Dao를 분리했던 이유는 ReservationService를 단일로 테스트할 수 있도록 Dao를 테스트 대역으로 대체할 수 있도록 하기 위함도 있고 Dao는 데이터베이스가 아닌 다른 곳에 저장할 수 있다고 생각이 들어 이렇게 인터페이스를 중간에 추가하였습니다.

seovalue replied:
좋네요!

changuii replied:
감사합니다.

changuii commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public ReservationResponse createReservation(final ReservationRequest reservationRequest) {
+        final Reservation notSavedReservation = new Reservation(
+                null,
+                reservationRequest.name(),
+                reservationRequest.date(),
+                null
+        );
```
여기서 null의 의미는 실제로 이 값이 아직 없어서 `Reservation`이라는 객체는 아직 미완성되어 있다는 것을 명확하게 표현하기 위해 이렇게 만들었습니다.

seovalue replied:
요 부분에서, 

timeId로 time을 미리 조회해온 다음, reservation을 만들고, 그걸 save 해도 될 것 같은데 어떻게 생각하시는지 궁금해요~!

changuii replied:
`reservation`을 저장할 때, time은 굳이 필요하지 않고 id만 있으면 된다고 생각했습니다.
- 먼저 timeId로 time을 조회하는 경우, 불필요하게 time까지 조회해야 하는데 굳이 이렇게 해주어야 하는 이유를 잘 모르겠습니다.
- 존재하지 않는 timeId를 reservation에 저장하려는 경우, 외래키 제약조건을 만족하지 않기 때문에 똑같이 예외가 발생합니다. 이 부분에 대한 예외처리를 하지 않았는데 추가하겠습니다!

seovalue replied:
무슨 의미인지 이해했어요.
그럼 reservation 도메인에서도 timeId만 관리하는게 명확하지 않을까요?
도메인에서는 reservationTime이 존재하는데, 처음에 생성은 null로 하게 된다는 것이 예약에서는 시간이 없어도 된다는 의미인가?로 여겨지는 것 같아요.

changuii replied:
우테코에 들어오기 전에 프로젝트를 할때는 "객체"에 대해서 많은 고민을 하지 않고 "스프링 부트"라는 툴로 서버를 만들어내는 느낌을 많이 받았어요
우테코를 들어오고 나서 레벨1을 하면서 객체지향에 많이 고민해보고 객체란 무엇일까에 대해 많이 고민을 해봤었고 그런 고민을 하고난 후 레벨2에서 다시 스프링을 학습하면서 많은 의문점들이 생겼어요
객체는 동작으로 정의되고 상태는 동작에 필요한 것들이 유기적으로 생겨난다고 생각해요! 그런데, 제가 지금 작성한 코드는 그렇게 되지 않는 것 같아요 아직 동작들이 없이 간단한 저장을 구현하고 있어서 그런지는 모르겠으나 `Reservation`이 객체지향에서 말하는 객체라기 보다는 데이터베이스 스키마에 맞춰 구현되어 있다고 느껴졌어요
실제 객체가 변하는 이유는 객체가 가지는 책임인 "동작"이 분리되거나 변경될 때 변해야 하지만 현재 `Reservation`은 그런 이유가 아니라 데이터베이스에 의존되어 데이터베이스 스키마의 변경에 따라 상태들이 변하고 있다고 생각해요 그렇다면 도메인 객체를 DAO에서 사용하는게 올바른 것일까?에 많은 의문이 들었어요 DAO에게 전달할 때도 데이터베이스에 전달하는 DTO가 있어야 하는게 아닐까?라는 생각도 했구요
조앤이 말씀하신 부분도 같은 맥락이라고 생각해요. 객체지향에서의 객체는 참조를 통해서 관계를 갖지만 데이터베이스 측면에서는 id를 이용한 외래키를 통해 관계를 맺고 있기 때문에 두 개념이 충돌하는 부분이라고 생각이 들었어요
그래서 Dao에서는 데이터베이스 테이블을 나타내는 객체를 만들어 사용해야하지 않을까? 라고 생각도 들었습니다.

하지만, 우테코 밖에서 공부를 했을 당시에도 그렇고 지금 와서도 이런 DTO의 존재는 본적이 없어요
- 그래서 전에 이것 저것 찾아보면서 헥사고날 아키텍처에 대해서 알게되었고 깊게 살펴보지는 않았지만 이 구조에서는 JPA entity가 도메인 계층 외부에 존재하면서 데이터를 저장할 때만 entity로 변환해주고 이 entity에는 실제 도메인 로직이 담겨있지는 않더라구요! 그래서 당시에는 이 고민에 대해서 고민해보다가 이 구조는 복잡하기 때문에 레이어드 아키텍처를 많이 사용하는구나 정도로만 결론을 지었었어요

지금 계속 고민해도 아직 로직이 데이터베이스에 저장, 삭제, 조회정도의 기능만 있어서 결론이 맺어지지 않는 것 같아서 레벨1 때처럼 더 복잡한 도메인 로직이 생기면 추가로 고민해볼 것 같습니다!

seovalue replied:
ㅎㅎ네네! 
암튼 제 의견을 정리해드리자면 저는 Reservation이 생기기 위해서는 시간이 필수적으로 필요한 요소라면 늘 할당되었음을 보장해야한다는 생각이긴 해요!!

(그리고 이건 별개인데ㅎㅎ 부기가 주시는 답변들에는 부기의 고민이 잘 녹아있는 것 같아요. 이게 꽤 좋은데 한편으로는 길다보니 정확히 요점이 뭔지 파악하기 어렵기도 해요. 그래서 섹션을 분리해서 작성해보셔도 좋을 것 같아요! ## 고민 ## 정리 요런식으로 나눠본다면 리뷰어도 좀 더 확실히 이해하고 의견을 전달드릴 수 있을 것 같아요!!)

changuii commented on src/main/java/roomescape/time/TimeController.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            timeService.deleteTimeById(id);
+            return ResponseEntity.ok().build();
+        } catch (final IllegalArgumentException e) {
+            return ResponseEntity.notFound().build();
+        }
```
Time과 Reservation을 삭제할 때, 삭제하려는 대상이 존재하지 않는 경우 응답으로 404 not found를 반환한 이유는 사용자가 요청한 url에 해당하는 "자원"인 "time" 혹은 "Reservation"이 not fount 즉, 존재하지 않는다를 명확하게 표현하고 싶었습니다.

seovalue replied:
not found를 반환했을 때 유저가 취할 수 있는 액션이 있나요? 어떤 의미가 있는건지 궁금해요

changuii replied:
사용자에게 삭제가 정상적으로 이루어지지 않았음을 알려주고, API 클라이언트 입장에서 요청을 나중에 다시 시도하거나 문의를 남겨달라고 alert를 띄워줄 수 있을 것 같습니다.
- 현재는 html과 js에 not found에 대한 액션이 없지만, 클라이언트 입장에서 구분이 안되면 구분 작업을 취할 수 없기 때문에 이렇게 처리하였습니다.

seovalue replied:
이해했어요! 제 의도는 not found에 한정해서 생각해보자면, 
존재하지 않는 리소스에 대한 삭제 요청이 들어온 것 자체가 불가능한 행위이기 때문에, 이에 대해서 굳이 alert을 줘야할까? 라는 생각이었어요!

changuii replied:
해당 자원에 대해서 여러 사람이 동시에 접근할 수 있으니 삭제할 때, 리소스가 존재하지 않는다면 "이미 삭제된 예약입니다."와 같은 알람을 줄 수는 있을 것 같긴 해요. 다만, 조앤이 말씀하신 것 처럼 굳이라고 생각한다면 그렇게 생각이 들긴 합니다!

changuii commented on src/test/java/roomescape/admin/AdminApiTest.java:
```diff
@@ -7,14 +7,20 @@
 
 @SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
-@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
```
여기서 `AFTER_EACH_TEST_METHOD`를 사용한 이유는 `BEFORE_EACH_TEST_METHOD`를 사용하였을 경우 다음과 같은 이슈가 발생했기 때문입니다.
1. `BEFORE_EACH_TEST_METHOD`를 사용하면 JUnit5가 테스트를 위해 테스트 클래스 인스턴스를 생성을 하고 난 후 애플리케이션을 종료시키고 다시 실행시킵니다.
2. 여기서 문제가 발생하는데, 저는 port와 JdbcTemplate와 같은 값들의 변경 가능성을 제거하고 싶었기 때문에 final 키워드를 붙였고 때문에 "생성자 주입" 방식을 사용하여 빈을 주입해주었습니다.
3. "생성자 주입"은 Junit5가 테스트 클래스 인스턴스를 생성하는 시점에 빈들이 주입이되었고 "1"번에서 말했던 것 처럼 애플리케이션이 종료되고 새로 실행되기 전에 빈들이 주입되었기 때문에 `DirtiesContext`가 종료시킨 애플리케이션 시점의 빈들이 주입되어 있는 상태가 되었습니다.

결론적으로 빈들이 주입되는 시점 때문에 예외가 발생해 이를 해결하고자 AFTER_EACH_TEST_METHOD를 사용했습니다.
- 이렇게 바꾸면 애플리케이션이 실행된 후, 테스트 클래스를 인스턴스화 하고, 때문에 이 시점에 빈들이 주입되어 예외가 발생하지 않았습니다.

seovalue replied:
해당 AdminApiTest에서는 DirtiesContext는 없어도 되는거 아닌가요?

changuii replied:
해당 부분에 대해서는 의식적으로 고민하지 못했던 것 같습니다. 피드백 반영하겠습니다. 감사합니다 :)

changuii commented on src/test/java/roomescape/reservation/FakeReservationDao.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Objects;
+import roomescape.time.Time;
+
+public class FakeReservationDao implements ReservationDao {
+
+    private static final Time DUMMY_TIME = new Time(1L, LocalTime.of(12, 40));
+
+    private Long NEXT_ID = 1L;
+    private final List<Reservation> reservations = new ArrayList<>();
+    private final List<Long> invokeDeleteId = new ArrayList<>();
```
DAO를 테스트 대역으로 대체하고 Service를 테스트하기 위해 `Fake` 객체를 만들어 사용했습니다.
- 다만, delete 같은 경우에는 Spy처럼 Dao의 delete 메서드가 호출되는 것만 검증하면 되었기에 `invokeDeleteId`를 활용해 호출 여부만 확인했습니다.

이렇게 한 이유는 service의 createReservation(), findAllReservation() 메서드들의 결과는 반환값이기 때문에 이렇게 메서드 호출을 검증하는 것은 중간 과정을 검증하는 것이기 때문에 호출을 검증하면 안된다고 생각합니다.
하지만, deleteReservationById()의 경우에는 dao의 deleteById가 메서드 기능의 "결과"라고 생각하여 이렇게 메서드 호출을 검증하게 되었습니다.

seovalue replied:
delete는 void이기 때문에 실제로 호출했는지 검증한다! 인거죠? 이해했어요.

FakeDao를 썼을 때 어떤 장/단점이 존재할까요?

changuii replied:
FakeDao를 사용했을 때의 장점은 외부 의존성을 끊어내고 우리가 테스트 하고자하는 코드를 테스트할 수 있는 것이라고 생각합니다. 단위 테스트가 외부 의존성에 의존하면 할수록 테스트마다 의존성을 관리해주고 신경써주어야 하는 리소스가 생긴다고 생각하고 외부 의존성(현재 프로세스 외부의 프로세스)와 통신하는 비용이 크기 때문에 테스트에 소요되는 시간이 많이 늘어난다고 생각해요 그렇게되면 결국 단위 테스트에 걸리는 시간이 많아지고 빠르게 피드백을 받지 못하는 문제가 발생한다고 생각합니다. 빠르게 피드백을 받지 못하면 단위 테스트를 많이 돌려보지 못하는 상황이 나오게 되고 단위 테스트의 장점이 퇴색될 수 있다고 생각합니다.

FakeDao를 사용했을 때의 단점은 외부 의존성에 대한 연결을 테스트해볼 수 없는 문제가 있습니다. 다만 이 단점은 통합 테스트나 엔드 투 엔드 테스트를 통해 검증해야하는 부분이라고 생각해요.
그리고 Fake도 코드기 때문에 관리해야할 코드가 증가하는 문제점이 있는 것 같아요

spy를 사용했을 때의 문제도 고민해보았습니다.
행동을 검증하는 것이기 때문에 실제 구현체에 대해 테스트가 결합되는 문제가 있었습니다. 실제로 구현체 외부의 클라이언트 입장에서는 어떻게 목적에 달성하는지는 테스트할 필요가 없고 목적을 잘 달성하기만 하면 되지만 현재는 "dao의 delete 메서드를 호출한다"라는 세부 구현내용까지 알고 있습니다. 때문에 리팩터링을 통해 세부 구현 내용이 변경되었을 때 목적은 올바르게 달성하지만, 세부 구현내용의 변경으로 테스트가 실패할 수 있는 문제점이 있습니다.
- 실제 동작은 올바르게 하지만, 테스트가 실패한다. 이런 문제가 계속 발생하면 개발자는 테스트를 믿지 못하게 될 수 있고 테스트를 리팩터링하는데 비용이 많이 소모되어 테스트를 유지보수하는데 드는 비용이 테스트의 가치보다 높아지게될 가능성이 있습니다!


결론적으로 저는 단위 테스트를 작성할 때는 최대한 외부 의존성을 제거할 것 같고 통합 테스트나 엔드 투 엔드 테스트를 작성할 때, 외부 의존성을 포함시킨 테스트를 할 것 같습니다.!

changuii commented on src/test/java/roomescape/reservation/ReservationApiTest.java:
```diff
@@ -45,14 +58,11 @@ void post() {
-                .then().log().all()
-                .statusCode(200)
-                .body("id", is(1));
+        // given
+        givenCreateTime();
+        givenCreateReservation();
 
+        // when & then
         RestAssured.given().port(port).log().all()
                 .when().get("/reservations")
                 .then().log().all()
```
조앤이 말씀하신 것처럼 RestAssured를 묶어내니 가독성이 훨씬 향상되었습니다.
다만, when & then에 해당하는 부분은 이전에 말씀드렸던 것과 동일하게 생각하여 검증하려는 목표는 그대로 묶어내지 않고 두었습니다!

seovalue replied:
반영해주셔서 감사해요!

changuii reviewed (commented):
제 의도를 작성해보았습니다.
리뷰 감사합니다 조앤 ! :)

seovalue commented on README.md:
```diff
@@ -81,4 +81,129 @@ DELETE /reservations/1 HTTP/1.1
+- API 명세를 따라 시간 관리 API를 구현하세요.
+- 페이지는 templates/admin/time.html 파일을 이용하세요.
+
+#### 시간 생성
```
👍 명세 굿

changuii replied:
감사합니다!

seovalue commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,36 +1,20 @@
 
 public record Reservation(
-        long id, String name, LocalDate date, LocalTime time
+        Long id, String name, LocalDate date, Time time
```
long 대신 Long으로 해주신 이유가 궁금해요~

changuii replied:
primitive type을 사용하는 경우 id를 0으로 표현하거나 특정 상수를 "빈 값"으로 나타내야 하지만, 이렇게 `Long` wrapper type을 사용하면 명확하게 id가 빈 값이라고 `null`을 통해 알려줄 수 있다고 생각해요!

seovalue commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,36 +1,20 @@
-
-        return new Reservation(writeId, name, date, time);
+    public Long id(){
+        Objects.requireNonNull(id);
+        return id;
```
👍

changuii replied:
감사합니다 :)

seovalue commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -1,40 +1,49 @@
+    private final ReservationService reservationService;
+
+    public ReservationController(
+            @Autowired final ReservationService reservationService
```
Autowired는 어떤 역할을 해요?

changuii replied:
Autowired 어노테이션은 ApplicationContext에서 관리하고 있는 Bean을 주입해달라고 알려주는 역할을 합니다.
- 생성자 주입에서는 이 어노테이션을 생략해도 된다고 나와있지만 실제로 빈을 주입하고 있음을 명확하게 하기 위해 그대로 달아주었습니다.

seovalue commented on src/main/java/roomescape/reservation/ReservationJdbcDao.java:
```diff
@@ -0,0 +1,82 @@
+            @Autowired JdbcTemplate jdbcTemplate
+    ) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
```
simpleJdbcInsert는 어떤 장점이 있어서 사용하는 거예요?

changuii replied:
simpleJdbcInsert를 사용하면, 값들을 map에 담아서 보낼 수 있는데, 현재 방식이 기존에 사용하던 `preparedStatement`보다 훨씬 가독성이 좋다고 생각합니다. 특히, `preparedState`는 인덱스기반 파라미터를 지원하기 때문에 이를 신경써야 하기도 합니다.
그리고 고수준의 추상화된 기능을 사용하기 위해 JdbcTemplate를 사용하고 있는데 id값을 반환받기 위해 connection과 prepareStatement를 사용하는게 어색하고, 일관성이 떨어지는 것 같다고 생각합니다.
추가로, keyHolder를 사용할 때는 코드가 이해하기 힘든데 반해, SimpleJdbcInsert는 직관적인 메서드 네이밍, `executeAndReturnKey()`처럼 코드가 훨씬 이해하기 쉽다고 생각합니다.

seovalue commented on src/main/java/roomescape/reservation/ReservationJdbcDao.java:
```diff
@@ -0,0 +1,82 @@
+
+    private static void validateUpdateSuccess(final int updatedCount) {
+        if (updatedCount == 0) {
+            throw new IllegalArgumentException("[ERROR]");
```
요거 메시지에 ERROR만 보내면 어떤 에러인지 의미 파악이 어렵지 않을까요?

changuii replied:
같은 백엔드 개발자인지 API 클라이언트인지 말씀이 모호해서 두 부분에 대해서 다 작성했습니다!
###### 협업하는 개발자
코드를 작성하면서 했던 생각은 자바에서 제공하는 `IllegalArgumentException`과 메서드 네이밍인 `validateUpdateSuccess`를 보면 충분히 argument에 문제가 존재하고, 성공적으로 업데이트를 수행하지 못했구나!라고 생각할 수 있지 않을까?라고 이해가 가능할 것이라 판단했습니다. 
좀 더 생각해보니 `[ERROR]` 이렇게 표현해서 애매모호하게 나타내지 않고 확실하게 문자열로 메시지를 적는게 훨씬 이해하기 쉽다고 생각이 들었습니다. (굳이 이유를 적지 않을 필요가 없다.)
###### API 클라이언트
API 클라이언트의 입장에서는 문제가 없다고 생각했습니다. API 클라이언트와 결과에 대한 정보를 나타내기 위해 HTTP status code로 충분히 소통하고 있기 때문입니다.!

말씀하신 부분 반영하겠습니다! 감사합니다 :)

seovalue replied:
클라이언트 사이드에서도 같은 status code 로 상황에 따라 여러 에러가 올 수 있다면, 충분히 message나 혹은 부가적인 정보로 분기를 태울 수 있을 것 같아요!

changuii replied:
맞아요 ! message에 대해서 부가적인 정보로 분기를 할 수 있을 것 같긴 합니다. 다만, API들에 대해서 실패에 대한 많은 정보를 제공했을 때 보안에 취약해지지는 않을까? 고민이 들었어요. 여러 open API들을 활용해보면서 api response message가 대부분 그렇게 친절하지는 않더라구요 그런 것들을 보면서 저는 이게 보안에 취약해지기 때문에 message를 그렇게 자세하게 적지 않는걸까? 라는 생각을 했어요
- api에 대한 실패 정보를 많이 노출했을 때 악의적인 공격자에게 너무 많은 정보를 주는 건 아닐까요? 

아직 어느 정도로 message를 보내야할지에 대한 기준이 없어서 모호하긴 합니다!
- https://swagger.io/blog/problem-details-rfc9457-doing-api-errors-well/
- 그래도 오류에 대한 표준이 존재하더라구요! 일단 현재는 이정도 depth까지만 학습하고 나중에 프로젝트를 하면서 기준을 세울 것 같습니다!

seovalue replied:
좋네요ㅎㅎ 나중에 프론트엔드와 협업하면서 다양한 케이스를 만나보며 경험해도 좋을 것 같아요!

seovalue commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,42 @@
+        final Long savedReservationId = reservationDAO.saveReservation(notSavedReservation,
+                reservationRequest.timeId());
+        final Reservation savedReservation = reservationDAO.findReservationById(savedReservationId);
+        return ReservationResponse.createResponse(savedReservation);
```
POSTMapping으로 생성된 객체는 생성된 리소스의 ID만을 담은 URI를 보내기도 하는 것 같은데, 객체 전체를 응답해주신 이유가 궁금해요!

changuii replied:
API 명세에 작성된대로 객체 전체를 응답해주도록 했습니다. 
API 명세가 주어진 것 자체가, 클라이언트에게 이렇게 개발할 것이라고 약속을 해두고 개발한다고 생각합니다.
현재는 클라이언트가 없지만 만약 프론트엔드와 협업하는 과정이라면, 프론트엔드와의 대화 없이 명세된 API를 바꾸면 소통에 오류가 많이 발생할 것이라고 생각합니다.
실제로 우테코에 들어오기 전에 프로젝트를 할때 이런 소통 오류가 많이 발생했었습니다. 그래서 최대한 API 명세 그대로 맞추려고 했습니다.

seovalue commented on src/main/resources/banner.txt:
```diff
@@ -0,0 +1,13 @@
+    ___          ___           ___           ___           ___           ___
```
ㅋㅋㅋㅋㅋㅋㅋ

seovalue commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE reservation_time
```
테이블은 reservation_time인데 도메인은 time으로 명명해주신 이유가 궁금해욧

changuii replied:
도메인 클래스 이름을 time으로 지은 이유는 데이터베이스 테이블을 고려하기 보다는 기존에 time으로 사용하고 있었기 때문에 이렇게 명명하였습니다.
- 현재는 time이라는 값이 Reservation에만 존재하기 때문에 문제가 없을 것 같지만, 추후 time이라는 값은 많은 곳에서 사용할 수 있는 값 객체로써 느껴지기 때문에 구분짓겠습니다.!!

seovalue commented on src/test/java/roomescape/database/DataSourceTest.java:
```diff
@@ -0,0 +1,146 @@
+import roomescape.reservation.dto.ReservationResponse;
+
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
```
DirtiesContext 없이는 해결할 수 없나요?

changuii replied:
고민을 쭉 해보았는데, 굳이 DirtiesContext를 사용해서 컨텍스트를 종료하고 실행하지 않고 실제로 변경되는 부분만 초기화를 해주면, 더 빠르게 테스트를 진행할 수 있을 것이라 판단했습니다.
그래서 실제로 변경되는 부분인 데이터베이스만 초기화하여 테스트 속도를 향상시켰습니다.!
- 기존에 약 2.6s가 걸리던 시간을 모든 DirtiesContext를 제거하고 약 2.05초로 감소했습니다.

seovalue reviewed (changes requested):
안녕하세요 부기! 조앤이에요ㅎㅎ 
크게 수정할 건 없을 것 같은데 이번에도 의견이 궁금한 것들만 몇가지 남겨봤어요~!~!
의견 많이 주세요!ㅎㅎ

changuii committed:
refactor: Exception message를 자세히 작성

changuii committed:
feat: TimeId가 존재하지 않을 때 발생하는 예외를 잡아서 BadRequest를 반환하도록 예외 처리

changuii committed:
test: 불필요한 DirtiesContext를 제거

- 불필요하게 컨텍스트를 종료하고 실행하는데 소비되는 시간을 줄이기위해 제거

changuii committed:
refactor: Time을 ReservationTime으로 네이밍

- Time은 범용적으로 사용되어질 수 있는 네이밍이기 때문에 어디서 사용되어지는지 명확하게하기 위해 변경

changuii committed:
test: DirtiesContext 사용을 최소화

changuii committed:
test: 존재하지 않는 시간으로 예약 생성 테스트

changuii committed:
test: DirtiesContext 제거

- DirtiesContext를 제거하여 약 500ms 테스트 소요 시간 감소 (약 20% 감소)
- 불필요한 객체 생성 제거 (fakeDao 및 service)

changuii committed:
refactor: 불필요한 static 키워드 제거

changuii reviewed (commented):
리뷰 감사합니다 조앤 :)

changuii commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -1,40 +1,54 @@
     ) {
-        final Reservation savedReservation = reservations.save(reservation);
-        return ResponseEntity.ok(savedReservation);
+        try {
+            final ReservationResponse response = reservationService.createReservation(request);
+            return ResponseEntity.ok(response);
+        } catch (final DataIntegrityViolationException e){
+            return ResponseEntity.badRequest().build();
+        }
```
실제로 외래키 제약조건을 일부러 어겨서 어떤 예외가 발생하는지 실험을 해보았습니다.
- 이때, JdbcTemplate가 예외를 추상화하여 던지기 때문에 저는 스프링에서 추상화시킨 예외가 발생할 것이라 기대했습니다.

하지만, 실험을 해보았을 때는, h2에서 던지는 예외인 `JdbcSQLIntegrityConstraintViolationException`가 발생했습니다.
그래서 저는 가설을 세워보았습니다.
-  SimpleJdbcInsert가 JdbcTemplate와 다른 로직을 가지고 처리하는 것일까?

그래서 코드를 뜯어보니, SimpleJdbcInsert도 JdbcTemplate.execute()를 통해 쿼리를 보내고 있었습니다.
이해가 안되서 실제로 JdbcTemplate가 예외를 포장하고 있는지 확인하기 위해 중단점을 찍고 디버깅을 해보았습니다.
- 확인 결과 JdbcTemplate.execute()내부 `translateException()` 메서드에서 정상적으로 `DataIntegritiViolationException`으로 포장해주고 있었고, `DataIntegrityViolationException`을 catch로 잡아서 printStackTrace()를 호출하여 확인하였을 때는 정상적으로 포장이 되어있었습니다.

그렇다면 어디서 스프링이 추상화한 예외를 없애버리는걸까? 를 고민하면서 크루들과 이야기를 통해 유추를 해보았습니다.
- 실제로 예외를 포장하고 있다면, 우리가 제어하는 영역 앞부분인 Spring MVC, Tomcat 쪽에서 발생하는 것이 아닐까?라는 의견이 나왔습니다.

그래서 중단점을 찍고 디버깅을 해본 결과 Spring MVC를 지나, Tomcat영역에서 `StandarWrapperValve` 클래스의 `invoke` 메서드에서 `StandardWrapper.getRootCause(e)`를 통해 실제 Cause를 가진 Exception을 추출하고 있었습니다.

결과적으로 이런 로직이 있는 이유를 가추해보았을 때, Spring 프레임워크에서 제공한 추상화된 예외를 포장하여 던져주었지만 개발자가 이 예외에 대한 처리를 해주지 않고 Tomcat까지 왔기 때문에 불필요한 포장된 예외를 벗겨내고 정확히 발생한 이유에 해당하는 예외만 던져주기 위해 이런 로직이 있다고 추측했습니다.

seovalue replied:
오우 뒷단까지 확인해보셨군요!
h2말고 mysql을 사용하면 어떤 예외가 발생돼요?

그리고 추가적으로 해당 try-catch가 Controller에서 관리되는 것에 대해서 어떻게 생각해요?

changuii replied:
###### mysql로 변경했을 때 발견한 차이점들
- mysql을 쓰니 auto table create가 작동하지 않는다. [스프링 공식 문서](https://docs.spring.io/spring-boot/how-to/data-initialization.html)에서 확인해보니 기본적으로 스프링이 내장 인메모리 데이터베이스를 사용하는 경우에만 동작한다고 나와있었다. 이를 활성화해주기 위해 `spring.sql.init.mode` 옵션을 always로 설정하여 해결할 수 있었습니다.
- mysql은 테이블에 대한 대소문자를 구분한다. [mysql 공식 문서](https://dev.mysql.com/doc/refman/8.4/en/identifier-case-sensitivity.html)에서 확인해보니, 운영체제에 따라 대소문자를 구분한다고 나와있었습니다. 제가 사용하는 MacOS는 기본적으로 대소문자를 구분하지 않고, Unix 기반의 운영체제는 대소문자를 구분한다고 나와있었습니다. 제가 도커 컨테이너에서 mysql을 실행했고 각 도커 컨테이너가 운영체제를 가지지는 않지만 기본적으로 호스트의 리눅스 커널을 사용한다고 하여 이를 컨테이너 내부 터미널에서 확인해보니 `Linux aa72997ac3f6 5.15.49-linuxkit-pr #1 SMP PREEMPT Thu May 25 07:27:39 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux` 이렇게 리눅스로 나와서 mysql이 대소문자 구분을 한 것 같습니다.
- mysql과 문법이 다소 다른점이 존재합니다. h2에서 auto_incement 값을 초기화해주는 쿼리가 mysql과 약간 달랐습니다. sql을 직접 사용하다보니 각 데이터베이스 방언에 맞는 쿼리로 제가 직접 바꿔주어야 했습니다.

###### 발생한 예외의 차이점
```java
java.sql.SQLIntegrityConstraintViolationException: Cannot add or update a child row: a foreign key constraint fails (`mydatabase`.`reservation`, CONSTRAINT `reservation_ibfk_1` FOREIGN KEY (`time_id`) REFERENCES `reservation_time` (`id`))
```
- mysql에서는 java.sql 패키지에 정의된 예외가 발생했습니다.
- 확인해보니 h2에서 던지는 예외는 mysql에서 던지는 예외와 동일한`SQLIntegrityConstraintViolationException`을 상속받은 서브 클래스였다.
h2가 자신만의 예외를 만들어서 던지는 이유는 사실 잘 모르겠다. 추측상으로 아래의 내용들을 예외에 더 담아서 던지기 위해서가 아닐까?라고 생각이 들었습니다.
```java
private final String originalMessage;  
private final String stackTrace;  
private String message;  
private String sql;
```

###### try-catch가 controller에서 관리되는 것에 대한 생각
Controller에서 try-catch를 통한 분기 처리를 하는 건 자연스럽다고 느껴졌습니다. controller의 역할은 API 요청을 받아서 다음 계층으로 호출하고, 호출에 대한 응답을 만들어주는 것이라고 생각합니다. 때문에 잘못된 처리에 대한 응답도 controller에서 분기 처리를 통해서 해주는게 저는 자연스럽다고 느껴졌습니다.
다만, 같은 예외에 대해서 다른 controller들이 각각 예외 처리를 해주어야 하기 때문에 중복 코드가 많이 발생하는 문제가 존재하긴 했습니다.
- 이는 나중에 `ControllerAdvisor`를 사용해서 묶어내면 쉽게 관리할 수 있을 것이라 생각했습니다!

seovalue replied:
오웅 많이 시도해보셨네요ㅎㅎ
관점의 차이인 것 같은데, 제 생각에 DB와 관련된 에러는 컨트롤러보다는 가장 가까운 Dao(Repository) 혹은 Service에서 처리하는게 어떨까 싶어요.
컨트롤러는 사실 클라이언트와 소통을 하는 창구라고 생각하는데, DB의 에러까지 끌고오는게 너무 많은 걸 알게되는 게 아닌가 싶은 생각이 들어서 질문 남겨봤어요!

changuii replied:
> DB의 에러까지 끌고오는게 너무 많은 걸 알게되는 게 아닌가 싶은 생각이 들어서 질문 남겨봤어요!

말씀하신 방향으로는 생각을 못해보았던 것 같습니다. 감사합니다 :)

changuii committed:
fix: refactor 기능으로 잘못 수정된 부분을 다시 수정

changuii commented on src/test/java/roomescape/reservation/FakeReservationDao.java:
```diff
@@ -0,0 +1,61 @@
+                .anyMatch(value -> Objects.equals(value, id));
+    }
+
+    public void clear(){
+        NEXT_ID = 1L;
+        reservations.clear();
+        invokeDeleteId.clear();
+    }
```
fake dao에 clear()메서드를 추가하여 객체 내부 상태를 초기화하는 로직을 만들었습니다.
- 이를 통해 테스트에서 초기화하기 위해 매번 객체를 새로 생성해주어야 하는 불필요한 과정을 제거하였습니다.

changuii commented on src/test/java/roomescape/reservation/ReservationApiTest.java:
```diff
@@ -7,32 +7,56 @@
+        TIME_BODY.put("startAt", "10:00");
+    }
+
+    @BeforeEach
+    void setUp() {
+        jdbcTemplate.update("DELETE FROM RESERVATION");
+        jdbcTemplate.update("DELETE FROM RESERVATION_TIME");
+        jdbcTemplate.update("ALTER TABLE RESERVATION ALTER COLUMN id RESTART WITH 1");
+        jdbcTemplate.update("ALTER TABLE RESERVATION_TIME ALTER COLUMN id RESTART WITH 1");
     }
```
실제로 애플리케이션에서 변경되어 다른 테스트에 영향을 주는 부분인 database를 초기화하는 로직을 추가하고, `@DirtiesContext`는 제거하였습니다.

changuii reviewed (commented):
의도와 학습 과정에서 발생한 고민들을 작성하였습니다!

seovalue commented on src/test/java/roomescape/reservation/ReservationApiTest.java:
```diff
@@ -55,11 +66,27 @@ void post() {
+    @Test
+    void post2() {
+        // given
+        givenCreateReservationTime();
+        givenDeleteReservationTime();
```
Long createReservationTimeId = givenCreateReservationTime();
givenDeleteReservationTime(createReservationTimeId)

이런식으로 되어도 좋을 것 같네요~

changuii replied:
처음에는 이렇게도 해볼까라고 고민해보고 적용했었지만 일관된 데이터를 생성하고, 삭제하는게 더 테스트를 읽기 쉽지 않을까? 라고 생각해서 현재처럼 바꿨습니다.
- 좀 더 고민해보니 현재는 그런 테스트가 없지만 5개의 데이터를 생성하고, 삭제한다거나 하는 상황에서는 계속 메서드를 생성할 수는 없으니 인자로 받아서 처리하는게 더 나을 것 같습니다. 피드백 감사합니다 :)

seovalue reviewed (approved):
안녕하세요 부기~
이번 단계는 이정도에서 마무리해도 될 것 같아요!
다만 이전에 남겨주셨던 질문에 대한 코멘트 남겨뒀는데, 혹시라도 질문 있으시면 추가로 남겨주세요~!

changuii reviewed (commented):
생각해볼 수 있는 부분들을 많이 리뷰해주셔서 감사합니다 조앤 ! :) 😃
